File: scripts/dgrsdt.ts
import { chromium } from 'playwright';

const CATEGORY_B_SUBCATEGORIES = [
  "ABDC", "De_Gruyter", "Erih_plus", "Journal_quality",
  "AERES", "CNRS", "SCOPUS", "Finacial_Times"
];

async function getJournalInfo(journalName: string) {
  const browser = await chromium.launch({ headless: false });
  const page = await browser.newPage();

  // === CATEGORY A ===
  console.log(`🔎 Searching Category A ➜ ${journalName}`);
  try {
    await page.goto("https://www.dgrsdt.dz/fr/revues_A", { timeout: 60000 });
    await page.waitForSelector("input.input-search-job", { timeout: 30000 });

    const searchInput = await page.$("input.input-search-job");
    if (searchInput) {
      await searchInput.fill(journalName);
      await page.waitForTimeout(2000); // Let DOM update

      const rows = await page.$$("li.table-row");
      for (const row of rows) {
        const text = await row.innerText();
        if (text.toLowerCase().includes(journalName.toLowerCase())) {
          const cols = await row.$$("div.col");
          await printJournal(cols);
          await browser.close();
          return;
        }
      }
    }
  } catch {
    console.log("⚠️ Failed to process Category A");
  }

  // === CATEGORY B ===
  console.log(`🔎 Searching Category B ➜ ${journalName}`);
  await page.goto("https://www.dgrsdt.dz/fr/revues_B", { timeout: 60000 });

  for (const subcat of CATEGORY_B_SUBCATEGORIES) {
    console.log(`🔄 Trying subcategory ➜ ${subcat}`);
    try {
      const button = await page.$(`button#${subcat}`);
      if (!button) continue;

      await button.click();
      await page.waitForSelector("input.input-search-job", { timeout: 5000 });

      const searchInput = await page.$("input.input-search-job");
      if (searchInput) {
        await searchInput.fill(journalName);
        await page.waitForTimeout(2000);

        const rows = await page.$$("li.table-row");
        for (const row of rows) {
          const text = await row.innerText();
          if (text.toLowerCase().includes(journalName.toLowerCase())) {
            const cols = await row.$$("div.col");
            await printJournal(cols, subcat);
            await browser.close();
            return;
          }
        }
      }
    } catch (e) {
      console.log(`⚠️ Failed to search in subcategory ${subcat} ➜ ${e}`);
    }
  }

  console.log(`❌ Journal not found in Category A or B`);
  await browser.close();
}

async function printJournal(cols: any[], subcat?: string) {
  console.log("\n📘 Journal:", await cols[0].innerText());
  console.log("🏢 Publisher:", await cols[1].innerText());
  console.log("📚 ISSN:", await cols[2].innerText());
  console.log("🌐 EISSN:", await cols[3].innerText());
  if (subcat) {
    console.log("🏷️ Subcategory:", subcat);
  }
}

// test
const journal = "3D Printing and Additive Manufacturing";
getJournalInfo(journal);

File: tests/dgrsdt.py
from playwright.sync_api import sync_playwright
import time
from difflib import SequenceMatcher

CATEGORY_B_SUBCATEGORIES = [
    "ABDC", "De_Gruyter", "Erih_plus", "Journal_quality",
    "AERES", "CNRS", "SCOPUS", "Finacial_Times"
]

def is_similar(a, b, threshold=0.85):
    return SequenceMatcher(None, a.lower(), b.lower()).ratio() >= threshold

def get_journal_info(journal_name: str):
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=False)
        page = browser.new_page()

        # === CATEGORY A ===
        print(f"🔎 Searching Category A ➜ {journal_name}")
        page.goto("https://www.dgrsdt.dz/fr/revues_A", timeout=60000)
        try:
            page.wait_for_selector("input.input-search-job", timeout=30000)
            search_input = page.query_selector("input.input-search-job")
            search_input.fill(journal_name)
            time.sleep(2)  # Let DOM update

            rows = page.query_selector_all("ul.responsive-table li.table-row")
            for row in rows:
                title_div = row.query_selector('div[data-label="Journal_Title"]')
                if title_div and is_similar(journal_name, title_div.inner_text()):
                    cols = row.query_selector_all("div.col")
                    print_journal(cols)
                    browser.close()
                    return
        except Exception as e:
            print(f"⚠️ Failed to process Category A ➜ {e}")

        # === CATEGORY B ===
        print(f"🔎 Searching Category B ➜ {journal_name}")
        page.goto("https://www.dgrsdt.dz/fr/revues_B", timeout=60000)

        for subcat in CATEGORY_B_SUBCATEGORIES:
            print(f"🔄 Trying subcategory ➜ {subcat}")
            try:
                button = page.query_selector(f"button#{subcat}")
                if not button:
                    continue
                button.click()
                page.wait_for_selector("input.input-search-job", timeout=5000)
                search_input = page.query_selector("input.input-search-job")
                search_input.fill(journal_name)
                time.sleep(2)

                rows = page.query_selector_all("ul.responsive-table li.table-row")
                for row in rows:
                    title_div = row.query_selector('div[data-label="Journal_Title"]')
                    if title_div and is_similar(journal_name, title_div.inner_text()):
                        cols = row.query_selector_all("div.col")
                        print_journal(cols, subcat)
                        browser.close()
                        return
            except Exception as e:
                print(f"⚠️ Failed to search in subcategory {subcat} ➜ {e}")

        print(f"❌ Journal not found in Category A or B")
        browser.close()

def print_journal(cols, subcat=None):
    print("\n📘 Journal:", cols[0].inner_text())
    print("🏢 Publisher:", cols[1].inner_text())
    print("📚 ISSN:", cols[2].inner_text())
    print("🌐 EISSN:", cols[3].inner_text())
    if subcat:
        print("🏷️ Subcategory:", subcat)

# 
journal = "19th-Century Music"
get_journal_info(journal)
